### PR TITLE
Build docker image

### PR DESCRIPTION
**Pull request by AlexanderWert**
Adds a step in the CI to create a Docker image for the agent

### CODE REVIEW

File: .ci/Makefile [hunk 1]
```diff
@@ -18,6 +21,14 @@ clean:
 env:
 	env
 
+# Create Docker image
+build-docker: validate-branch-name
+	docker build -t $(DOCKER_REGISTRY)/$(DOCKER_IMAGE_NAME):$(BRANCH_NAME) \
+  --build-arg AGENT_DIR=$(DIST_DIR) ..
+
+push-docker: build-docker
+	../dev-utils/push-docker.sh $(DOCKER_REGISTRY) $(DOCKER_IMAGE_NAME) $(BRANCH_NAME)
```
Comment (trentm) [inline .ci/Makefile:30]: Forgot to 'git add dev-utils/push-docker.sh'?

File: .ci/Makefile [hunk 2]
```diff
@@ -18,6 +21,14 @@ clean:
 env:
 	env
 
+# Create Docker image
+build-docker: validate-branch-name
```
Comment (felixbarny) [inline .ci/Makefile:25]: Is `validate-branch-name` a copy/paste leftover?

File: .ci/Makefile [hunk 3]
```diff
@@ -18,6 +21,14 @@ clean:
...
+	docker build -t $(DOCKER_REGISTRY)/$(DOCKER_IMAGE_NAME):$(BRANCH_NAME) \
```
Comment (felixbarny) [inline .ci/Makefile:26]: Does the BRANCH_NAME include the `v` prefix, as in `v3.1.2`? If so, we probably want to remove it.

Comment (felixbarny) [reply .ci/Makefile:26]: Probably beyond the scope of this but we might also want to publish a `latest` tag and a tag for the latest major version, such as `3.x`. Not that `latest` doesn't necessarily refer to the most recent release, but the one that has the highest semver version number. For example `3.1.0 > 3.0.1`, even if `3.0.1` has been released after `3.1.0`.

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,7 @@
+# Pin to Alpine 3.9
+# For a complete list of hashes, see:
+# https://github.com/docker-library/repo-info/tree/master/repos/alpine/remote
+FROM alpine@sha256:115731bab0862031b44766733890091c17924f9b7781b79997f5f163be262178
+RUN mkdir /usr/agent
+ARG AGENT_DIR
+COPY ${AGENT_DIR} /usr/agent/nodejs
```
Comment (trentm) [inline Dockerfile:7]: For comparison the Java APM agent docker container is using "/usr/agent/elastic-apm-handler{,.jar}". With this the Node.js APM agent will be at "/usr/agent/nodejs/node_modules/elastic-apm-node/...", I think (I haven't run this build yet :).

**Review by trentm — COMMENTED** [review]
Thanks!

File: .ci/Makefile (hunk 1 above)
Comment (AlexanderWert) [reply .ci/Makefile:30]: yes, forgot it, added now

File: .ci/Makefile (hunk 2 above)
Comment (trentm) [reply .ci/Makefile:25]: Nope, that's good to use. It is defined below and ensures the `$(BRANCH_NAME)` var is set for this target.

File: .ci/Makefile (hunk 3 above)
Comment (trentm) [reply .ci/Makefile:26]: In current usage the `BRANCH_NAME` does *not* include the `v` prefix.

File: .ci/Makefile [hunk 5]
```diff
@@ -1,13 +1,16 @@
 SHELL = /bin/bash -eo pipefail
 
 AWS_FOLDER = ../build/aws
-LAMBDA_LAYER_ZIP_FILE = ../build/lambda-layer-zip/elastic-apm-node-lambda-layer.zip
+DIST_DIR = /build/dist/nodejs
```
[the remark this answers is not in the document]
Comment (AlexanderWert) [reply .ci/Makefile:4]: In Dockerfile it's always relative to the Dockerfile.
I first had the Dockerfile in the .ci sub dir and tried with relative `../build/dist/nodejs` but that does not work because all things added in the Dockerfile need to be on the same or below level in the directory structure relative to the Dockerfile

File: .ci/Makefile (hunk 3 above)
Comment (trentm) [reply .ci/Makefile:26]: +1 to "latest". We won't have much need for a "3.x" tag until/if we do a "4.x" major version of the APM agent.

Comment (AlexanderWert) [reply .ci/Makefile:26]: > Does the BRANCH_NAME include the v prefix, as in v3.1.2? If so, we probably want to remove it.

Yes, it does. Will try to remove it.

> Probably beyond the scope of this but we might also want to publish a latest tag ...

I added pushing of the `latest` tag already in `dev-utils/push-docker.sh`, but without the advanced logic, yet. 
Do you know how we can query the tags on a docker repo (e.g. from something like this https://container-library.elastic.co/r/observability/apm-agent-java ) to implement this advanced logic?

Comment (AlexanderWert) [reply .ci/Makefile:26]: | In current usage the BRANCH_NAME does not include the v prefix.

@trentm With my local testing it did. But I guess it behaves different on CI?

Comment (trentm) [reply .ci/Makefile:26]: If it helps with the logic, I don't think we've *ever* done the equivalent of a "3.0.1" release after having already released "3.1.0" or later. I.e. I don't think this code needs to worry about whether the current release *is* the "latest". Perhaps just have a "IS_LATEST" var hardcoded to true. If/when we have a "3.x" maintenance branch and "main" is for version 4.x of the agent, then we can set `IS_LATEST=false` for the 3.x branch (or just delete the block of code that adds the "latest" tag).

File: Dockerfile (hunk 4 above)
Comment (trentm) [reply Dockerfile:7]: I don't know exactly how the k8s attachment will work.  I'm guessing it will mount one of these directories in a target container and add a `NODE_OPTIONS="--require $mountPath"` envvar.

I *think* that would work without the `/node_modules/` component in the path. In other words, I think it might be possible to have the agent Docker image use "/usr/agent/elastic-apm-node/..." for its path, if you want.

Comment (trentm) [reply Dockerfile:7]: https://github.com/elastic/apm-agent-nodejs/issues/2622#issuecomment-1131573905 mentions perhaps wanting this Docker image to be also usable for container-based Lambda functions. That might dictate what path we want the APM agent to be installed into.

File: .ci/Makefile (hunk 3 above)
Comment (felixbarny) [reply .ci/Makefile]: > Do you know how we can query the tags on a docker repo (e.g. from something like this https://container-library.elastic.co/r/observability/apm-agent-java ) to implement this advanced logic?

This script returns the latest version number:
```
git tag --list "v*" | cut -c2- | sort --version-sort | tail -n 1
```

You could just check whether the current version is the same.

Comment (felixbarny) [reply .ci/Makefile]: Another alternative would be `git tag --list --sort=version:refname "v*" | cut -c2- | tail -n1`

If the tags may contain suffix that indicates pre-releases, such as `1.2.3-pre1`, or `1.2.3.RC1` (the latter is used in the Java agent) we can use this: `git -c 'versionsort.suffix=.RC' tag --list --sort=version:refname "v*" | cut -c2- | tail -n1`

File: Dockerfile (hunk 4 above)
Comment (astorm) [reply Dockerfile:7]: TL;DR; -- the path the agent's copied to in the container is arbitrary and its path doesn't need a particular name.  

Re: k8s attachment

> I don't know exactly how the k8s attachment will work. I'm guessing it will mount one of these directories in a target container and add a NODE_OPTIONS="--require $mountPath" envvar.

That's almost it.  The additional nuance is that the agent's source will be copied out of the Docker container and into a mounted volume at a fixed path.  The command that currently does this copying is here: https://github.com/elastic/apm-mutating-webhook/blob/2c6a4cecf33fb8022eb1f3a365e6705345ff908a/patch.go#L150.  (part of the work of K8s attachment is changing this copy to be able to work with multiple files instead of a single file `jar`) The some `NODE_OPTIONS` and `NODE_PATH` setting will happen in order to load the agent from this spot.

Comment (trentm) [reply Dockerfile:7]: Thanks.
We coudl consider *just* setting/updating `NODE_OPTIONS` and not bothering with `NODE_PATH`, right?

Comment (astorm) [reply Dockerfile:7]: > We could consider just setting/updating NODE_OPTIONS and not bothering with NODE_PATH, right?

That would be ideal -- but without a `NODE_PATH` set (it doesn't seem like?) manual instrumentation in the user's code would able to find the agent module.  Is that incorrect?

**Review by astorm — COMMENTED** [review]
Thanks for putting this together @AlexanderWert -- overall it looks great but I have few questions for you (or maybe for @v1v) that's more about me not knowing how our jenkins systems work. 

Context: I tried building the container (without pushing) locally with the following invocation.

    % cd .ci
    % DOCKER_IMAGE_NAME=observability/apm-agent-nodejs DOCKER_REGISTRY='docker.elastic.co' BRANCH_NAME='v3.34.0' make build-docker

However, the docker build failed.
    
    docker build -t docker.elastic.co/observability/apm-agent-nodejs:3.34.0 \
      --build-arg AGENT_DIR=/build/dist/nodejs ..
    [+] Building 2.1s (7/7) FINISHED                                                                                                                                     
     => [internal] load build definition from Dockerfile                                                                                                            0.0s
     => => transferring dockerfile: 335B                                                                                                                            0.0s
     => [internal] load .dockerignore                                                                                                                               0.0s
     => => transferring context: 2B                                                                                                                                 0.0s
     => [internal] load metadata for docker.io/library/alpine@sha256:115731bab0862031b44766733890091c17924f9b7781b79997f5f163be262178                               1.9s
     => [1/3] FROM docker.io/library/alpine@sha256:115731bab0862031b44766733890091c17924f9b7781b79997f5f163be262178                                                 0.0s
     => => resolve docker.io/library/alpine@sha256:115731bab0862031b44766733890091c17924f9b7781b79997f5f163be262178                                                 0.0s
     => => sha256:115731bab0862031b44766733890091c17924f9b7781b79997f5f163be262178 1.64kB / 1.64kB                                                                  0.0s
     => => sha256:ab3fe83c0696e3f565c9b4a734ec309ae9bd0d74c192de4590fd6dc2ef717815 528B / 528B                                                                      0.0s
     => => sha256:82f67be598ebc8d968137c18521fe174ca6afc9b542aa5773c32b3a3970e647c 1.51kB / 1.51kB                                                                  0.0s
     => [internal] load build context                                                                                                                               0.0s
     => => transferring context: 27B                                                                                                                                0.0s
     => CACHED [2/3] RUN mkdir /usr/agent                                                                                                                           0.0s
     => ERROR [3/3] COPY /build/dist/nodejs /usr/agent/nodejs                                                                                                       0.0s
    ------
     > [3/3] COPY /build/dist/nodejs /usr/agent/nodejs:
    ------
    failed to compute cache key: "/build/dist/nodejs" not found: not found
    make: *** [build-docker] Error 1

It seemed to be trying to copy files from `/build/dist/nodejs` into the container but no such files existed.  I also tried running `BRANCH_NAME='v3.34.0' make dist` to make the distribution and it was built into `../build/dist` and not `/build/dist/nodejs`.  I was finally able to get things built locally by running

    DOCKER_IMAGE_NAME=observability/apm-agent-nodejs DOCKER_REGISTRY='docker.elastic.co' BRANCH_NAME='v3.34.0' DEST_DIR='/Users/astorm/Documents/github/elastic/apm-agent-nodejs/dist/nodejs' make build-docker

In other words -- by running `make dist` and then pointing `DEST_DIR` at the correct build directory I was able to get this working.

All of this raised two questions in my mind:

1. Is there any quick way to make this build locally in one step while maintaining the needed functionality in our Jenkins environments?

2. Does the `stage('Push Docker Image')` section need to run `make dist` as well, or can we safely assume that the source will be there for this stage (this is a very naive "how does our Jenkins work" question)

File: Dockerfile (hunk 4 above)
Comment (trentm) [reply Dockerfile:7]: > manual instrumentation in the user's code would able to find the agent module. Is that incorrect?

You are correct. I think there are some finer points around that that might be easier to discuss on a call (or better on https://github.com/elastic/apm-agent-nodejs/issues/2622). Basically whether we *want* the user to be able to `require('elastic-apm-node')` and get this externally installed package. My (limited) understanding of the OTel equivalent is that they are not modifying `NODE_PATH`.  If the programmer is adding manual instrumentation then they are likely better served having their own `"elastic-apm-node"` in their package.json dependencies.

**Comment by AlexanderWert** [comment]
@astorm 
> 1. Is there any quick way to make this build locally in one step while maintaining the needed functionality in our Jenkins environments?
> 2. Does the `stage('Push Docker Image')` section need to run `make dist` as well, or can we safely assume that the source will be there for this stage (this is a very naive "how does our Jenkins work" question)

To run it locally you need execute the following:
```
cd .ci
make dist && make push-docker
```

The `push-docker` stage assumes that the `make dist` stage has been executed before.
Our current `Publish to AWS` stage has the same assumption. 
In the Jenkinsfile those steps are executed in sequence as part of the `Release` stage.

````

stage('Release') {
      options { skipDefaultCheckout() }
      when {
        beforeAgent true
        tag pattern: 'v\\d+\\.\\d+\\.\\d+', comparator: 'REGEXP'
      }
      environment {
        SUFFIX_ARN_FILE = 'arn-file.md'
      }
      stages {
        stage('Dist') {
          steps {
            withGithubNotify(context: "Dist") {
              ...
              withNodeJSEnv(version: "${env.BUILD_NODE_VERSION}"){
                dir("${BASE_DIR}"){
                  cmd(label: 'make dist', script: 'make -C .ci dist')
                }
              }
            }
          }
        }
        stage('Push Docker Image') {
          steps {
            withGithubNotify(context: "Push-Docker-Image") {
              dir("${BASE_DIR}"){
                dockerLogin(secret: "${env.DOCKER_SECRET}", registry: "${env.DOCKER_REGISTRY}", role_id: 'apm-vault-role-id', secret_id: 'apm-vault-secret-id')
                cmd(label: 'make push-docker', script: 'make -C .ci push-docker')
              }
            }
          }
        }
        stage('Publish to AWS') {
          steps {
            withGithubNotify(context: "Publish") {
              withGoEnv(){
                withAWSEnv(secret: 'secret/observability-team/ci/service-account/apm-aws-lambda', forceInstallation: true, version: '2.4.10') {
                  dir("${BASE_DIR}"){
                    cmd(label: 'make publish-in-all-aws-regions', script: 'make -C .ci publish-in-all-aws-regions')
                    cmd(label: 'make create-arn-file', script: 'make -C .ci create-arn-file')
                  }
                }
              }
            }
          }
````

Does it answer your questions?

**Comment by astorm** [comment]
> Does it answer your questions?

Yup -- this part in particular

> The push-docker stage assumes that the make dist stage has been executed before. Our current Publish to AWS stage has the same assumption. In the Jenkinsfile those steps are executed in sequence as part of the Release stage.

I tend to try and make each build not rely on artifacts from previous steps -- but as long as there's prior art and the steps are sequential it sounds good to me. 👍
